[mlir] Add a ValueSemantics trait.

pifon2a opened the pull request:
We need to distinguish ShapedTypes with and without value semantics. This is needed for downstream users to define their custom vector and tensor
types that can work with the arith/math dialect.

RFC https://discourse.llvm.org/t/rfc-mlir-types-with-encoding/80189

joker-eph commented on mlir/include/mlir/IR/BuiltinTypes.h:
```diff
@@ -39,6 +39,11 @@ struct IntegerTypeStorage;
 struct TupleTypeStorage;
 } // namespace detail
 
+/// Type trait indicating that the type can be an operand to an elementwise op.
```
This description, as well as the name of the trait does not seem to be something that we couldn't apply to Memref. 
However this isn't the intent right?

pifon2a replied:
No, it is not the intent. How would you phrase it better? I want this trait to be for value types like vector or tensor, but not for memrefs. MemRefs are just pointers, it is not clear what we actually map.

joker-eph replied:
Memref are more than pointers: they have a shape, which means you can actually iterate all the elements inside a memref. This makes it a candidate for a "map" function I believe (I understand this isn't what you want, which is why I'm picking on the name).
I think @jpienaar proposed something for this?

pifon2a replied:
`MappableValueBasedContainer`?

jreiffers replied:
Isn't this conflating two different traits? For an `arith` op on a shaped type you want it to have value semantics and you want the right element type. Why does that need to be expressed in a single trait?

pifon2a replied:
Added the second trait.

joker-eph commented on mlir/include/mlir/IR/BuiltinTypes.h:
```diff
@@ -39,6 +39,16 @@ struct IntegerTypeStorage;
+/// Type trait indicating that the type can be an operand to an elementwise op.
+template <typename ConcreteType>
+class MappableContainer
+    : public TypeTrait::TraitBase<ConcreteType, MappableContainer> {};
```
Do we need this right now knowing we have the ShapedTypeInterface?

pifon2a replied:
We don't. Good catch, Mehdi!

ThomasRaoux commented:
looks like there are some build errors on the bot:
```

In file included from /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-tvbq4-1/llvm-project/github-pull-requests/flang/lib/Optimizer/CodeGen/CGOps.cpp:33:
--
  | /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-tvbq4-1/llvm-project/github-pull-requests/build/tools/flang/include/flang/Optimizer/CodeGen/CGOps.cpp.inc:43:62: error: use of undeclared identifier 'ValueSemantics'; did you mean 'mlir::ValueSemantics'?
  | if (!((((type.isSignlessIntOrIndex())) \|\| (((type.hasTrait<ValueSemantics>())) && ([](::mlir::Type elementType) { return (elementType.isSignlessIntOrIndex()); }(::llvm::cast<::mlir::ShapedType>(type).getElementType())))) \|\| ((type.isSignedInteger())) \|\| ((::llvm::isa<::fir::IntegerType>(type))))) {
  | ^~~~~~~~~~~~~~
  | mlir::ValueSemantics


```

pifon2a commented:
> looks like there are some build errors on the bot:
> 
> ```
> 
> In file included from /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-tvbq4-1/llvm-project/github-pull-requests/flang/lib/Optimizer/CodeGen/CGOps.cpp:33:
> --
>   | /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-tvbq4-1/llvm-project/github-pull-requests/build/tools/flang/include/flang/Optimizer/CodeGen/CGOps.cpp.inc:43:62: error: use of undeclared identifier 'ValueSemantics'; did you mean 'mlir::ValueSemantics'?
>   | if (!((((type.isSignlessIntOrIndex())) \|\| (((type.hasTrait<ValueSemantics>())) && ([](::mlir::Type elementType) { return (elementType.isSignlessIntOrIndex()); }(::llvm::cast<::mlir::ShapedType>(type).getElementType())))) \|\| ((type.isSignedInteger())) \|\| ((::llvm::isa<::fir::IntegerType>(type))))) {
>   | ^~~~~~~~~~~~~~
>   | mlir::ValueSemantics
> ```

Should work now. Let's see.

joker-eph reviewed (approved):
LGTM but if @jpienaar has also time to review, it'd be better.

ThomasRaoux reviewed (approved):
Looks good to me as well

jpienaar reviewed (approved):
Looks good to me, I think it reads simply and meaning is clear (I'm reminded we are missing doc generation for these, unrelated of course).

pifon2a committed:
[mlir] Add a ValueSemantics trait.

We need to distinguish ShapedTypes with and without value semantics. This is needed for downstream users to define their custom vector and tensor
types that can work with the arith/math dialect.

RFC https://discourse.llvm.org/t/rfc-mlir-types-with-encoding/80189

stellaraccident reviewed (commented):
Belated lgtm. Thanks.
